Readme.md

```diff
@@ -1,65 +1,118 @@
-                MyUploadControlValidationSettings.Settings, out errors, (s, e) => { },
-                UploadControl_FilesUploadComplete);
-```
-5. (*OPTIONAL*) To pass information about uploaded files to the client side, add the *FilesUploadComplete* handler and assign required info to e.CallbackData:
-```cs
-public void UploadControl_FilesUploadComplete(object sender, FilesUploadCompleteEventArgs e) {
-            UploadedFile[] files = ((MVCxUploadControl)sender).UploadedFiles;
-            // files processing...
-            e.CallbackData = "Info";
-}
-
-```
-6. (*OPTIONAL*) Get the information passed from your Controller to the client, in the client-side [*ASPxClientUploadControl.FilesUploadComplete*](https://documentation.devexpress.com/AspNet/DevExpress.Web.Scripts.ASPxClientUploadControl.FilesUploadComplete.event) event handler:
-```cs
- @Html.DevExpress().UploadControl(settings => {
-    settings.Name = "UploadControl";
-    ...
-    settings.ClientSideEvents.FilesUploadComplete = "OnFilesUploadComplete";}).GetHtml()
-```
-```js
-  function OnFilesUploadComplete(s, e) {
-        var data = e.callbackData; // info from your Controller
-        ...      
+        settings.ClientSideEvents.FilesUploadComplete = "OnFilesUploadComplete";
+    }).GetHtml()
+    ```
+
+5. Call the [Html.BeginForm](https://learn.microsoft.com/en-us/dotnet/api/system.web.mvc.html.formextensions.beginform?view=aspnet-mvc-5.2#system-web-mvc-html-formextensions-beginform(system-web-mvc-htmlhelper-system-string-system-string)) method overload to wrap the Upload Control in the `<form>` HTML tag. To set the tag's action attribute, pass a name of an action method to the `BeginForm` method:
+
+    ```razor
+    @using (Html.BeginForm("UploadControlUpload", "Home")) {
+        @Html.DevExpress().UploadControl(settings => {
+            // ...
+        }).GetHtml()
     }
-```
-### Result:
+    ```
+
+6. In the action method, call the Upload Control's [GetUploadedFiles](https://docs.devexpress.com/AspNetMvc/DevExpress.Web.Mvc.UploadControlExtension.-l------------y-) method overload to get the uploaded files and handle the server-side [FilesUploadComplete](https://docs.devexpress.com/AspNet/DevExpress.Web.ASPxUploadControl.FilesUploadComplete) event:
+
+    ```cs
+    public ActionResult UploadControlUpload() {
+        string[] errors;
+        UploadedFile[] files = UploadControlExtension.GetUploadedFiles("UploadControl",
+            MyUploadControlValidationSettings.Settings, out errors, (s, e) => { },
+            UploadControl_FilesUploadComplete);
+        return null;
+    }
+    ```
+7. In the `FilesUploadComplete` event handler, save the uploaded files on the server. Write names of uploaded files and their URLs in a string. Assign this string to the [CallbackData](https://docs.devexpress.com/AspNet/DevExpress.Web.FilesUploadCompleteEventArgs.CallbackData) event argument to pass the string to the client:
```

Comment:
> Write names of uploaded files and their URLs **in** a string.

My first question is addressed to the used preposition. "In" normally answers the aspect of place. Where? In a string. "To", respectively, conveys the aspect of target. Where? To a string.

This case urges us to use a combined preposition "into". But, since we decided not to use it, we have to choose the most important aspect. And, in this particular case, the most important aspect is the aspect of target. Because the string is just a container for your information.

I understand that your code example below describes how to do it, yet still would have probably considered adding some information on "how to write names of uploaded files and their URLs"? In what format? What should follow what? What separator characters should one insert between entries within a string? Etc., etc.

I do not insist, but at least consider this...

```suggestion
7. In the `FilesUploadComplete` event handler, save the uploaded files on the server. Write names of uploaded files and their URLs to a string. Assign this string to the [CallbackData](https://docs.devexpress.com/AspNet/DevExpress.Web.FilesUploadCompleteEventArgs.CallbackData) event argument to pass the string to the client:
```

Reply:
I do not think we should tell our users how exactly they should pass information to the client to implement this particular use case. 

As I see, the main goal of the example is to illustrate the basic concepts of working with the Upload Control. And it doesn't really matter what we send from the server to the client in this example, it's important that it can be done. If the reader is really interested in this particular use case, he or she can review the code example below.

Anyway, thank you for your comment. I appreciate your desire to improve our learning materials.

Reply:
Alright, I understand your point. Focus is important, purpose is important. Thank you for your critical attitude. It's invaluable.